funfact/lang/_semiring.py

```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+from ._ast import Primitives as P
+from ._tsrex import TsrEx, EinopEx
+
```

Comment:
We can add support for log-spaced Einsum contractions.
We need  to add a reduction of the form shown below to the `DummyBackend` in `_einop.py` and create
an `EinopEx` for this.

```
def logspaceSum(data, axis=None):
    return np.log(np.sum(np.exp(data), axis=axis))
```